subsys/llext/llext.c

```diff
@@ -0,0 +1,470 @@
+							       sizeof(name),
+							       K_NO_WAIT);
+			strncpy(ext->sym_tab.syms[j].name, name, name_sz);
+			ext->sym_tab.syms[j].addr =
+				(void *)((uintptr_t)ext->mem[ldr->sect_map[sym.st_shndx]]
+					 + sym.st_value);
+			LOG_DBG("function symbol %d name %s addr %p",
+				j, name, ext->sym_tab.syms[j].addr);
+			j++;
+		}
+	}
+
+	/* relocations */
+	uintptr_t loc = 0;
+
+	pos = ldr->hdr.e_shoff;
+
+	for (i = 0; i < ldr->hdr.e_shnum - 1; i++) {
+		llext_seek(ldr, pos);
+		llext_read(ldr, &shdr, sizeof(elf_shdr_t));
+
+		pos += ldr->hdr.e_shentsize;
+
+		/* find relocation sections */
+		if (shdr.sh_type != SHT_REL && shdr.sh_type != SHT_RELA) {
+			continue;
+		}
+
+		rel_cnt = shdr.sh_size / sizeof(elf_rel_t);
+
+
+		llext_seek(ldr, ldr->sects[LLEXT_SECT_SHSTRTAB].sh_offset + shdr.sh_name);
+		llext_read(ldr, name, sizeof(name));
+
+		if (strncmp(name, ".rel.text", sizeof(name)) == 0 ||
+		    strncmp(name, ".rela.text", sizeof(name)) == 0) {
+			loc = (uintptr_t)ext->mem[LLEXT_MEM_TEXT];
+		} else if (strncmp(name, ".rel.bss", sizeof(name)) == 0) {
+			loc = (uintptr_t)ext->mem[LLEXT_MEM_BSS];
+		} else if (strncmp(name, ".rel.rodata", sizeof(name)) == 0) {
+			loc = (uintptr_t)ext->mem[LLEXT_MEM_RODATA];
+		} else if (strncmp(name, ".rel.data", sizeof(name)) == 0) {
+			loc = (uintptr_t)ext->mem[LLEXT_MEM_DATA];
+		}
+
+		LOG_DBG("relocation section %s (%d) linked to section %d has %d relocations",
+			name, i, shdr.sh_link, rel_cnt);
+
+		for (j = 0; j < rel_cnt; j++) {
+			/* get each relocation entry */
+			llext_seek(ldr, shdr.sh_offset + j * sizeof(elf_rel_t));
+			llext_read(ldr, &rel, sizeof(elf_rel_t));
+
+			/* get corresponding symbol */
+			llext_seek(ldr, ldr->sects[LLEXT_SECT_SYMTAB].sh_offset
+				    + ELF_R_SYM(rel.r_info) * sizeof(elf_sym_t));
+			llext_read(ldr, &sym, sizeof(elf_sym_t));
+
+			llext_seek(ldr, ldr->sects[LLEXT_SECT_STRTAB].sh_offset +
+				    sym.st_name);
+			llext_read(ldr, name, sizeof(name));
+
+			LOG_DBG("relocation %d:%d info %x (type %d, sym %d) offset %d sym_name "
+				"%s sym_type %d sym_bind %d sym_ndx %d",
+				i, j, rel.r_info, ELF_R_TYPE(rel.r_info), ELF_R_SYM(rel.r_info),
+				rel.r_offset, name, ELF_ST_TYPE(sym.st_info),
+				ELF_ST_BIND(sym.st_info), sym.st_shndx);
+
+			uintptr_t link_addr, op_loc, op_code;
+
+			/* If symbol is undefined, then we need to look it up */
+			if (sym.st_shndx == SHN_UNDEF) {
+				link_addr = (uintptr_t)llext_find_sym(&SYMTAB, name);
+
+				if (link_addr == 0) {
+					LOG_ERR("Undefined symbol with no entry in "
+						"symbol table %s, offset %d, link section %d",
+						name, rel.r_offset, shdr.sh_link);
+					/* TODO cleanup and leave */
+					continue;
+				} else {
+					op_code = (uintptr_t)(loc + rel.r_offset);
+
+					LOG_INF("found symbol %s at 0x%lx, updating op code 0x%lx",
+						name, link_addr, op_code);
+				}
+			} else if (ELF_ST_TYPE(sym.st_info) == STT_SECTION) {
+				link_addr = (uintptr_t)ext->mem[ldr->sect_map[sym.st_shndx]];
```

Comment:
I think link address should be:
`link_addr = (uintptr_t)ext->mem[ldr->sect_map[sym.st_shndx]] + *pop_loc + sym.st_value;`

with `pop_loc = (uint32_t*)op_loc;`

Reply:
This is determining the absolute linear address of the section to relocate, I'm not sure why it would be relative to the op code location in this case.

In the relocations in the hello world elf we can see the location of the string hello world is in .rodata and if this was working incorrectly the test would not be outputting hello world.

Reply:
Hi,

If you test with a second printk and a different string argument, a second relocation record is added with same value but string are differents.

Reply:
@florentaumaitre indeed you are right, working on a fix for this, will credit you, thank you!